File: Plugins/Orimath.Basics/ViewModels/EffectListViewModel.fs
namespace Orimath.Basics.ViewModels
open Mvvm
open Orimath.Plugins

type EffectListViewModel(workspace: IWorkspace, dispatcher: IDispatcher) =
    inherit NotifyPropertyChanged()
    let mutable effects = None
    member __.Effects =
        match effects with
        | Some(result) -> result
        | None ->
            let result =
                workspace.Effects
                |> Seq.map(fun e -> EffectViewModel(e, dispatcher))
                |> Seq.toArray
            effects <- Some(result)
            result

File: Plugins/Orimath.Basics.View/ViewModels/EffectListViewModel.cs
using System.Linq;
using Mvvm;
using Orimath.Plugins;

namespace Orimath.Basics.View.ViewModels
{
    public class EffectListViewModel : NotifyPropertyChanged
    {
        private readonly IWorkspace _workspace;
        private readonly IMessenger _messenger;

        private EffectViewModel[]? _effects;
        public EffectViewModel[] Effects => _effects ??= GetEffects();

        public EffectListViewModel(IWorkspace workspace, IMessenger messenger)
        {
            _workspace = workspace;
            _messenger = messenger;
        }

        private EffectViewModel[] GetEffects()
        {
            return _workspace.Effects
                .Select(e => new EffectViewModel(e, _messenger))
                .ToArray();
        }
    }
}
